### ISSUE TITLE
Gradient computation

### ISSUE THREAD

**Issue by Atcold** [jhjin/triplet-criterion#3]
``` lua
require 'nn'
require 'tripletCriterion'

-- Options                                       
batchSize = 20                                   
embSize = 4                                      
samples = 3                                      
blocks = 5                                       

emb = torch.randn(batchSize, embSize)            
t = torch.Tensor(batchSize)                      
for i = 0, blocks - 1 do                         
   for j = 1, samples do                         
      t[i*samples + j] = i + 1                   
   end                                           
end                                              
t[{ {samples*blocks, batchSize} }]:random(blocks)

crit = nn.TripletCriterion(samples, blocks)      
loss = crit:forward(emb, t)                      
gradInput = crit:backward(emb, t)                
```

The `gradInput` does not match the `input` (`emb`) dimensionality.

``` lua
== User Variables ==                        
[t]              = DoubleTensor - size: 20  
[blocks]         = 5                        
[embSize]        = 4                        
[samples]        = 3                        
[gradInput]      = DoubleTensor - size: 40x4
[loss]           = 0.0058705684636134       
[crit]           = nn.TripletCriterion      
[emb]            = DoubleTensor - size: 20x4
[batchSize]      = 20                       
```

I left some comments in the code. Please, have a look.

**Comment by jhjin** [comment]
Yes agree that there's a dimension mismatch. Here's a missing part from the old training script. Something like this should be placed in the package, not in the training script.

``` lua
local nb_backwards = math.max(1, opt.nSamples-1)
model:backward(inputs, gradOutputs:view(nb_backwards, opt.batchSize, gradOutputs:size(2)):sum(1):squeeze())
```

**Comment by Atcold** [comment]
Oh, so you send in batches of batches? Is it? Can you explain?
